toolz/itertoolz.py

```diff
@@ -1051,3 +1053,56 @@ def random_sample(prob, seq, random_state=None):
 
         random_state = Random(random_state)
     return filter(lambda _: random_state.random() < prob, seq)
+
+
+def _default_descend(x):
+    return not isinstance(x, (str, bytes, Mapping))
+
+
+def flatten(level, seq, descend=_default_descend):
+    """ Flatten a possibly nested sequence
+
+    Inspired by Javascript's Array.flat(), this is a recursive,
+    depth limited flattening generator. A level 0 flattening will
+    yield the input sequence unchanged. A level -1 flattening
+    will flatten all possible levels of nesting.
+
+    >>> list(flatten(0, [1, [2], [[3]]]))    # flatten 0 levels
+    [1, [2], [[3]]]
+    >>> list(flatten(1, [1, [2], [[3]]]))    # flatten 1 level
+    [1, 2, [3]]
+    >>> list(flatten(2, [1, [2], [[3]]]))
+    [1, 2, 3]
+    >>> list(flatten(-1, [1, [[[[2]]]]]))    # flatten all levels
+    [1, 2]
+
+    An optional ``descend`` function can be provided by the user
+    to determine which iterable objects to recurse into. This function
+    should return a boolean with True meaning it is permissible to descend
+    another level of recursion. The recursion limit of the Python interpreter
+    is the ultimate bounding factor on depth. By default, stings, bytes,
+    and mappings are exempted.
+
+    >>> list(flatten(-1, ['abc', [{'a': 2}, [b'123']]]))
+    ['abc', {'a': 2}, b'123']
+
+    See also:
+        concat
+     """
+    if level < -1:
+        raise ValueError("Level must be >= -1")
+    if not callable(descend):
```

Comment:
Maybe let users pass `None` for `descend` to say "I want the default value"?

```suggestion
    if descend is None:
        descend = _default_descend
    if not callable(descend):
```

This is something I almost always recommend for default arguments, especially when the default value is private, like it is here.

Because

1. It can help code that *programmatically* selects the `descend` argument.

2. It can help with writing wrappers of `flatten` which themselves let users pass an optional `descend`.